src/components/room-upgrade-handler.ts

```diff
@@ -12,102 +12,119 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    migrateStoreEntries: boolean;
+    onRoomMigrated?: (oldRoomId: string, newRoomId: string) => Promise<void>|void;
+    migrateEntry?: (entry: RoomBridgeStoreEntry, newRoomId: string) => Promise<RoomBridgeStoreEntry>;
+}
 
 /**
  * Handles migration of rooms when a room upgrade is performed.
  */
-class RoomUpgradeHandler {
+export class RoomUpgradeHandler {
+    private waitingForInvite = new Map<string, string>(); // newRoomId: oldRoomId
     /**
      * @param {RoomUpgradeHandler~Options} opts
      * @param {Bridge} bridge The parent bridge.
      */
-    constructor(opts, bridge) {
+    constructor(private readonly opts: RoomUpgradeHandlerOpts, private readonly bridge: any) {
         if (opts.migrateGhosts !== false) {
             opts.migrateGhosts = true;
         }
         if (opts.migrateStoreEntries !== false) {
             opts.migrateStoreEntries = true;
         }
-        this._opts = opts;
-        this._bridge = bridge;
-        this._waitingForInvite = new Map(); //newRoomId: oldRoomId
     }
 
-    onTombstone(ev) {
+    // eslint-disable-next-line camelcase
+    public async onTombstone(ev: {sender: string, room_id: string, content: {replacement_room: string}}) {
         const movingTo = ev.content.replacement_room;
         log.info(`Got tombstone event for ${ev.room_id} -> ${movingTo}`);
-        const joinVia = new MatrixUser(ev.sender).domain;
+        const joinVia = new MatrixUser(ev.sender).host;
         // Try to join the new room.
-        return this._joinNewRoom(movingTo, [joinVia]).then((couldJoin) => {
+        try {
+            const couldJoin = await this.joinNewRoom(movingTo, [joinVia]);
             if (couldJoin) {
-                return this._onJoinedNewRoom(ev.room_id, movingTo);
+                return this.onJoinedNewRoom(ev.room_id, movingTo);
             }
-            this._waitingForInvite.set(movingTo, ev.room_id);
+            this.waitingForInvite.set(movingTo, ev.room_id);
             return true;
-        }).catch((err) => {
+        }
+        catch (err) {
             log.error("Couldn't handle room upgrade: ", err);
             return false;
-        });
+        }
     }
 
-    _joinNewRoom(newRoomId, joinVia=[]) {
-        const intent = this._bridge.getIntent();
-        return intent.join(newRoomId, [joinVia]).then(() => {
+    private async joinNewRoom(newRoomId: string, joinVia: string[] = []) {
+        const intent = this.bridge.getIntent() as Intent;
+        try {
+            await intent.join(newRoomId, joinVia);
             return true;
-        }).catch((ex) => {
+        }
+        catch (ex) {
             if (ex.errcode === "M_FORBIDDEN") {
                 return false;
             }
             throw Error("Failed to handle upgrade");
-        })
+        }
     }
 
-    onInvite(ev) {
-        if (!this._waitingForInvite.has(ev.room_id)) {
+    // eslint-disable-next-line camelcase
+    public async onInvite(ev: {room_id: string}) {
```

Comment:
Why is this returning a boolean and what's the information this boolean carries?

Reply:
Have improved the JSDoc for the whole class.